### PR TITLE
Updated Cadence spec

### PR DESCRIPTION
**Pull request by worryg0d**
Done:
- added `primaryTargetCadence` to the cadence spec
- added `secondaryTargetCadence` to the cadence resource status 

To use the multi-region feature a customer should provide non-overlapping CIDR for the secondary mode cluster. Otherwise, the secondary mode cluster never gets into a running state.

### CODE REVIEW

File: apis/clusters/v1beta1/cadence_types.go [hunk 1]
```diff
@@ -305,6 +313,17 @@ func (cd *CadenceDataCentre) privateLinkToInstAPI() (iPrivateLink []*models.Priv
...
 }
 
+func (cs *CadenceSpec) TargetCadenceToInstAPI() []*models.TargetCadence {
+	var targets []*models.TargetCadence
+	for _, target := range cs.TargetPrimaryCadence {
+		targets = append(targets, &models.TargetCadence{
+			DependencyCDCID:   target.DependencyCDCID,
+			DependencyVPCType: target.DependencyVPCType,
+		})
+	}
```
Comment (tengu-alt) [inline apis/clusters/v1beta1/cadence_types.go:323]: Add a new line here and fix similar cases below please

Comment (worryg0d) [reply apis/clusters/v1beta1/cadence_types.go:323]: Fixed.

File: apis/clusters/v1beta1/cadence_types.go [hunk 2]
```diff
@@ -554,6 +590,28 @@ func (cs *CadenceSpec) validatePackagedProvisioning(old []*PackagedProvisioning)
...
+		}
+
+		if !ok {
+			return fmt.Errorf("targetPrimaryCadence is immutable")
+		}
+	}
+
+	return nil
+}
+
```
[the remark this answers is not in the document]
Comment (worryg0d) [reply apis/clusters/v1beta1/cadence_types.go]: It makes sense. Added validation on Cadence creation.
Also if there only 1 element exists then we can check like that:
```if *cs.TargetPrimaryCadence[1] != *old.TargetPrimaryCadence[1] { return err }``` .
Done.

**Review by tengu-alt — APPROVED** [review]

File: controllers/clusters/cadence_controller.go [hunk 3]
```diff
@@ -1179,6 +1191,24 @@ func (r *CadenceReconciler) updateDescriptionAndTwoFactorDelete(cadence *v1beta1
 	return nil
 }
 
+func areSecondaryCadenceTargetsEqual(first, second []*v1beta1.TargetCadence) bool {
```
Comment (ribaraka) [inline controllers/clusters/cadence_controller.go:1194]: first -> k8sTargets
second -> iTargets

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: controllers/clusters/cadence_controller.go (hunk 3 above)
Comment (worryg0d) [reply controllers/clusters/cadence_controller.go]: Renamed

File: controllers/clusters/cadence_controller.go [hunk 4]
```diff
@@ -874,6 +876,16 @@ func (r *CadenceReconciler) newWatchStatusJob(cadence *v1beta1.Cadence) schedule
...
+
+			if !cadenceTargetsEqual {
+				patch := cadence.NewPatch()
+				cadence.Status.TargetSecondaryCadence = iCadence.Status.TargetSecondaryCadence
+				err = r.Status().Patch(context.Background(), cadence, patch)
+				if err != nil {
+					l.Error(err, "Cannot patch cadence status with new target secondary cadence")
+					return err
+				}
+			}
```
[the remark this answers is not in the document]
Comment (worryg0d) [reply controllers/clusters/cadence_controller.go]: Fixed.

**Review by ribaraka — APPROVED** [review]

**Review by testisnullus — APPROVED** [review]

**Review by testisnullus — APPROVED** [review]

**Review by taaraora — APPROVED** [review]

**Pull request merged**